### PR TITLE
Fix #681 - Save all files before publishing anything

### LINKED ISSUE #681: Publishing doesn't publish the same version of the project as in the editor & preview window
Steps to reproduce
* Start a new project
* Make a small text change to the Carry On poster
* Immediately hit Publish
* Follow the link
* The changes you made are not reflected in the published version

Why? Probably because we don't auto-save except every 30 seconds and when you change files in the editor. 

Solution: Fire the auto-save call right before publishing to make sure we get the latest changes.

### PR DESCRIPTION
**Pull request by gideonthomas**

### CODE REVIEW

File: public/friendlycode/js/fc/filesystem-sync.js [hunk 1]
```diff
@@ -7,6 +7,12 @@ define(["jquery"], function($) {
     }
   }
 
+  function trigger(callbacks, args) {
```
Comment (humphd) [inline public/friendlycode/js/fc/filesystem-sync.js]: `triggerCallbacks`

File: public/friendlycode/js/fc/filesystem-sync.js [hunk 2]
```diff
@@ -19,29 +25,30 @@ define(["jquery"], function($) {
...
       request.always(function() {
         context.queueLength--;
         hideFileState(context.queueLength);
+        trigger(context._callbacks.afterEach, [error, path]);
       });
     }
 
     fs.readFile(path, function(err, data) {
       if(err) {
+        context.queueLength--;
```
Comment (humphd) [inline public/friendlycode/js/fc/filesystem-sync.js]: Why only in the error case?

Comment (gideonthomas) [inline public/friendlycode/js/fc/filesystem-sync.js]: it happens in the success case as well, just not immediately!

File: public/friendlycode/js/fc/filesystem-sync.js [hunk 3]
```diff
@@ -103,22 +114,124 @@ define(["jquery"], function($) {
...
+      if(fsync.queueLength > 0) {
+        return;
+      }
+
+      resetAfterEach(maybeContinue);
+      bramble.off("fileChange", fsync.handlers.change);
+      bramble.on("fileChange", cacheFilesSaved);
+
+      bramble.saveAll(function() {
+        bramble.off("fileChange", cacheFilesSaved);
```
Comment (humphd) [inline public/friendlycode/js/fc/filesystem-sync.js]: Is there a risk of losing track of file saves during the time that the `cacheFilesSaved` callback is off *and* the `fsync.handlers.change` callback is also not wired up yet (i.e., between now and when `finish()` gets called?  Do we risk any loss of info in that period if a save occurs?

File: public/friendlycode/js/fc/filesystem-sync.js [hunk 4]
```diff
@@ -89,6 +96,10 @@ define(["jquery"], function($) {
 
   function FSync() {
     this.queueLength = 0;
+    this._callbacks = {
+      beforeEach: [],
```
Comment (humphd) [inline public/friendlycode/js/fc/filesystem-sync.js]: Document what these are before and after.  It's a bit unclear.

File: public/friendlycode/js/fc/filesystem-sync.js [hunk 5]
```diff
@@ -103,22 +114,124 @@ define(["jquery"], function($) {
...
+        return;
+      }
+
+      fsync.afterEach(maybeFinish);
+      filesSaved.forEach(function(path) {
+        fsync.handlers.change(path);
+      });
+    }
+
+    function cacheFilesSaved(path) {
```
Comment (humphd) [inline public/friendlycode/js/fc/filesystem-sync.js]: I wonder if this name should indicate that you're caching a file *path* and not the file itself, so there's no confusion.

File: public/friendlycode/js/fc/filesystem-sync.js (hunk 3 above)
Comment (gideonthomas) [inline public/friendlycode/js/fc/filesystem-sync.js]: Well, one (and right now the only) scenario that I can imagine is if you click on "Publish", and while it is still publishing, you make changes in the editor (which will be auto-saved locally) and then once it's published, you do not make any more changes. Instead, you shutdown your browser, in which case it will be stored in IndexedDB but not persisted and will probably be overwritten (with whats on the server) when you open up your project again. **BUT** this is a very unlikely edge-case that will happen only if you time it correctly (stop changing anything as soon as publishing has ended) and want to unintuitively focus out of the Publish dialog box while publishing is occuring.

tldr; I don't think we should worry about this

File: public/friendlycode/js/fc/filesystem-sync.js [hunk 6]
```diff
@@ -103,22 +114,124 @@ define(["jquery"], function($) {
...
+      fsync.afterEach(maybeContinue);
+    } else {
+      maybeContinue();
+    }
+  };
+
+  // Add a callback to execute before every sync
+  // Each callback receives a `path` as an argument
+  FSync.prototype.beforeEach = function(callback) {
+    if(typeof callback !== "function") {
```
Comment (humphd) [inline public/friendlycode/js/fc/filesystem-sync.js]: I don't think this check is necessary; same is true below for `afterEach`.  I might change the names, though: `FSync.prototype.add{before|after}EachCallback`

File: public/friendlycode/js/fc/publisher.js [hunk 7]
```diff
@@ -63,6 +63,7 @@ define(function(require) {
   Publisher.prototype.init = function(project) {
     var publisher = this;
     var interface = publisher.interface;
```
Comment (humphd) [inline public/friendlycode/js/fc/publisher.js]: `interface` is a reserved word in JS.

**Comment by humphd** [comment]
I've done a round on the code, it's looking good minus my comments.  @sedge, since you know publish better than me, can you try this too?

**Comment by sedge** [comment]
Code looks good - testing now.

**Comment by sedge** [comment]
@gideonthomas walked me through it. r+
